internal/discovery/service.go

```diff
@@ -0,0 +1,140 @@
+package discovery
+
+import (
+	"context"
+	"fmt"
+	"github.com/flipkart-incubator/dkv/internal/hlc"
+	"github.com/flipkart-incubator/dkv/internal/master"
+	"github.com/flipkart-incubator/dkv/internal/storage"
+	"github.com/flipkart-incubator/dkv/pkg/ctl"
+	"github.com/flipkart-incubator/dkv/pkg/serverpb"
+	"gopkg.in/ini.v1"
+	"io"
+	"strconv"
+
+	"encoding/json"
+	"go.uber.org/zap"
+)
+
+/*
+This class contains the behaviour of receiving status updates from nodes in the cluster
+and providing the latest cluster info of active master / followers / slave of a region when requested
+*/
+
+type DiscoveryConfig struct {
+	// time in seconds after which the status entry of the region & node combination can be purged
+	StatusTTl uint64
+	// maximum time in seconds for the last status update to be considered valid
+	// after exceeding this time the region & node combination can be marked invalid
+	HeartbeatTimeout uint64
+}
+
+func NewDiscoverConfigFromIni(sect *ini.Section) (*DiscoveryConfig, error) {
+	sectConf := sect.KeysHash()
+	if statusTTL, err := strconv.ParseUint(sectConf["statusTTL"], 10, 64); err == nil {
+		if heartbeatTimeout, err := strconv.ParseUint(sectConf["heartbeatTimeout"], 10, 64); err == nil {
+			return &DiscoveryConfig{statusTTL, heartbeatTimeout}, nil
+		}
+	}
+	return nil, fmt.Errorf("Invalid discovery server configuration. Check section %s", sect.Name())
+}
+
+type discoverService struct {
+	// TODO - dkvServer should point to interface of master (currently distributedService)
+	dkvCli serverpb.DKVClient
+	logger *zap.Logger
+	config *DiscoveryConfig
+}
+
+// Create the dkv lookup key from region info
+// Since granularity of an update is at a region per node, the same is used for creating a key
+// As lookups will be mostly on database and possibly on vBucket (within a database) the same order is used for key
+// This enables faster scans for database/vBucket
+func createKeyToInsert(info *serverpb.RegionInfo) []byte {
+	return []byte(info.GetDatabase() + ":" + info.GetVBucket() + ":" + info.GetNodeAddress())
+}
+
+// Create lookup key based on input request
+// Using the sorted nature of keys, lookup by either database or by database and vBucket or all keys
+func createKeyToGet(request *serverpb.GetClusterInfoRequest) []byte {
+	if request.GetDatabase() != "" {
+		if request.GetVBucket() != "" {
+			return []byte(request.GetDatabase() + ":" + request.GetVBucket() + ":")
+		} else {
+			return []byte(request.GetDatabase() + ":")
+		}
+	} else {
+		return []byte{}
```

Comment:
@krishan1390 what is behaviour with empty byte? Underlying storage might not support this.

Reply:
empty byte is supposed to iterate over all keys. it works fine. test cases in TestDKVDiscoveryService calling GetClusterInfo without any vBucket or database

Reply:
But in that case the returned rows will be entire database. The next step where the value is expected to be in proper encoding would anyways fail for user data. So maybe this method should not return empty byte and instead return error?

Reply:
from a use case POV this is required because clients would want clusterinfo of all regions (databases and vbuckets) in the cluster. specifying filters would be unncessary client overhead. 

from a solution POV we will not return user data. it will return all rows of cluster info only. 

user data and cluster info are seperate regions.
currently multi region support in single node isn't enabled. so user data and cluster data are in different nodes.
when adding multi region support, the underlying dkvservice (and thus rocksdb store) will be isolated for each region. so each region can only access its own data.